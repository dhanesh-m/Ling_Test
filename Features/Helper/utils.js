// utils.js
export function searchByName(nameToSearch, dataObj) {
    for (const key in dataObj) {
      if (dataObj.hasOwnProperty(key)) {
        const entry = dataObj[key];
        if (
          entry.name.toLowerCase().trim() === nameToSearch.toLowerCase().trim()
        ) {
          return entry;
        }
      }
    }
    return null;
  }
  


  