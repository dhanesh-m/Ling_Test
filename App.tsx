import React, { useState} from 'react';
import {
  Alert,
  Keyboard,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import {dataObj} from './Data';
import AntDesign from 'react-native-vector-icons/AntDesign';
import {Table, Row, Rows} from 'react-native-table-component';
import {styles} from './styles';
import {User, AppProps, DataObj} from './types'; // Import the types
import { searchByName } from './utils';

const App: React.FC<AppProps> = () => {
  const [searchedUser, setSearchedUser] = useState<string>('');
  const [userList, setUserList] = useState<User[]>([]);
  const [errorMessage, setErrorMessage] = useState<string>('');

  const tableData = userList.map(item => [
    item.isSearchedUser ? (
      <Text style={styles.searchedText}>{item.name}</Text>
    ) : (
      item.name
    ),
    item.rank.toString(),
    item.bananas.toString(),
    item.isSearchedUser ? 'Yes' : 'No',
  ]);

  const handleSearch = () => {
    if (searchedUser.trim() != '') {
      const user = searchByName(searchedUser, dataObj);
      if (!user) {
        Alert.alert(
          'Alert Title',
          'This user name does not exist! Please specify an existing user name.',
          [
            {
              text: 'Cancel',
              onPress: () => console.log('Cancel Pressed'),
              style: 'cancel',
            },
            {text: 'OK', onPress: () => console.log('OK Pressed')},
          ],
        );
        setUserList([]);
      } else {
        const sortedUsers = Object.values(dataObj).sort(
          (a, b) => b.bananas - a.bananas,
        );
        const userIndex = sortedUsers.findIndex(u =>
          u.name.toLowerCase().includes(searchedUser.toLowerCase()),
        );
        if (userIndex < 0) {
          setErrorMessage('No users found');
          setTimeout(() => {
            setErrorMessage('');
          }, 200);
          setUserList([]);
        } else {
          const topUsers = sortedUsers.slice(0, 10).map((user, index) => ({
            name: user.name,
            rank: index + 1,
            bananas: user.bananas,
            isSearchedUser: user.name
              .toLowerCase()
              .includes(searchedUser.toLowerCase()),
          }));
          if (userIndex < 10) {
            setUserList(topUsers);
          } else {
            topUsers[9] = {
              name: user.name,
              rank: userIndex + 1,
              bananas: user.bananas,
              isSearchedUser: true,
            };
            setUserList(topUsers);
          }
          setErrorMessage('');
        }
      }
    } else {
      setErrorMessage('Please enter the name');
      setTimeout(() => {
        setErrorMessage('');
      }, 200);
    }
    Keyboard.dismiss();
  };

  return (
    <View style={styles.container}>
      <View style={styles.searchContainer}>
        <View style={styles.inputContainer}>
          <AntDesign name="search1" size={24} color="black" />
          <TextInput
            style={styles.input}
            placeholder="Enter name"
            onChangeText={text => setSearchedUser(text)}
            value={searchedUser}
          />
        </View>
        <TouchableOpacity
          onPress={() => handleSearch()}
          style={styles.buttonStyle}>
          <Text>Search</Text>
        </TouchableOpacity>
      </View>
      {errorMessage ? (
        <View style={styles.errorContainer}>
          <Text style={styles.errorText}>{errorMessage}</Text>
        </View>
      ) : (
        <View>
          {userList.length > 0 ? (
            <Table borderStyle={{borderWidth: 2, borderColor: '#c8e1ff'}}>
              <Row
                data={['Name', 'Rank', 'Bananas', 'Searched User']}
                style={styles.tableHeader}
                textStyle={styles.tableText}
              />
              <Rows data={tableData} textStyle={styles.tableText} />
            </Table>
          ) : null}
        </View>
      )}
    </View>
  );
};
export default App;

