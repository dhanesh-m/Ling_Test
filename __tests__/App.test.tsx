import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import App from '../App';

jest.mock('react-native-vector-icons/AntDesign', () => 'AntDesign');
jest.mock('../Features/Helper/utils', () => ({
  searchByName: jest.fn(),
}));

describe('App Component', () => {
  it('renders the component correctly', () => {
    const {getByText, getByPlaceholderText} = render(<App />);

    expect(getByText('Search')).toBeTruthy();
    expect(getByPlaceholderText('Enter name')).toBeTruthy();
  });

  it('handles search and displays results', async () => {
    const mockSearchByName = jest.fn(() => ({
      name: 'TestUser',
      rank: 1,
      bananas: 10,
    }));
    require('../Features/Helper/utils').searchByName = mockSearchByName;

    const {getByText, getByPlaceholderText, findByText} = render(<App />);

    const searchInput = getByPlaceholderText('Enter name');
    const searchButton = getByText('Search');
    fireEvent.changeText(searchInput, 'TestUser');
    fireEvent.press(searchButton);

    const resultName = findByText('TestUser');
    const resultRank = findByText('1');
    const resultBananas = findByText('10');

    expect(resultName).toBeTruthy();
    expect(resultRank).toBeTruthy();
    expect(resultBananas).toBeTruthy();
  });
});
