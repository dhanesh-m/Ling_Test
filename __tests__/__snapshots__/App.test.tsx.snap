// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Tests renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "paddingHorizontal": 15,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 20,
        "paddingTop": 30,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderColor": "#ccc",
          "borderRadius": 5,
          "borderWidth": 1,
          "flexDirection": "row",
          "paddingHorizontal": 10,
          "width": "80%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "black",
              "fontSize": 24,
            },
            undefined,
            {
              "fontFamily": "anticon",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        î™°
      </Text>
      <TextInput
        onChangeText={[Function]}
        placeholder="Enter name"
        style={
          {
            "color": "#333",
            "flex": 1,
            "height": 40,
          }
        }
        value=""
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "borderRadius": 5,
          "borderWidth": 1,
          "opacity": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 5,
        }
      }
    >
      <Text>
        Search
      </Text>
    </View>
  </View>
  <View />
</View>
`;
